home-assistant: handle packages with extra requirements

mweinelt opened the pull request:
###### Description of changes

https://github.com/NixOS/nixpkgs/pull/167560#issuecomment-1090978534

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
python3Packages.pygatt: split off GATTTOOL extra

mweinelt committed:
python3Packages.py3rijndael: init at 0.3.3

mweinelt committed:
python3Packages.samsungtvws: enable tests, split off extras

mweinelt committed:
python3Packages.samsungctl: split off websocket extra

mweinelt committed:
python3Packages.pytradfri: split off async extra

mweinelt committed:
python3Packages.adb-shell: split off extra requires

mweinelt committed:
python3Packages.androidtv: split off extra requires

mweinelt committed:
python3Packages.pure-python-adb: split off async requirements

mweinelt committed:
home-assistant: allow requirements parser to deal with extras

mweinelt committed:
home-assistant: update component packages

dotlambda commented on pkgs/development/python-modules/samsungctl/default.nix:
```diff
@@ -11,12 +15,17 @@ buildPythonPackage rec {
 
   # no tests
-  doCheck = false;
+  doCheck = true;
```
`checkPhase` should be specified or `pytestCheckHook` added to `checkInputs`
no need to specify `doCheck = true`

mweinelt replied:
Yeah, just leftovers from when I was checking out the tests.

dotlambda replied:
Shouldn't the dependencies be in propagatedBuildInputs if they're required for importing the module?

mweinelt replied:
Spot on.

dotlambda commented on pkgs/development/python-modules/androidtv/default.nix:
```diff
@@ -25,14 +25,24 @@ buildPythonPackage rec {
+    async = [
+      aiofiles
+    ];
+    usb = [
+      adb-shell.extras-require.usb
+    ];
```
```suggestion
    inherit (adb-shell.extras-require) usb;
```

dotlambda commented on pkgs/servers/home-assistant/component-packages.nix:
```diff
@@ -94,7 +94,7 @@
     "blueprint" = ps: with ps; [ ];
     "bluesound" = ps: with ps; [ xmltodict ];
-    "bluetooth_le_tracker" = ps: with ps; [ pygatt ];
+    "bluetooth_le_tracker" = ps: with ps; [ pygatt pygatt.extras-require.GATTTOOL ];
```
```suggestion
    "bluetooth_le_tracker" = ps: with ps; [ pygatt ] ++ pygatt.extras-require.GATTTOOL;
```
would be nice but I guess this is fine.

mweinelt replied:
Oh yeah, they're nested lists now. Shoot.

mweinelt replied:
Okay, I kinda went for a full reformat. PTAL.

mweinelt replied:
Merging in a bit so I can continue updating home-assistant (two minor releases) based on these changes.

mweinelt committed:
python310Packages.pure-python-adb: disable tests

All tests result in RuntimeError. Make the pythonImportsCheck for the
async parts conditional on doCheck, as checkInput only works when
doCheck is true.

mweinelt commented:
Result of `nixpkgs-review pr 167630` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>ghost</li>
    <li>home-assistant-component-tests.androidtv</li>
    <li>home-assistant-component-tests.bluetooth_le_tracker</li>
    <li>home-assistant-component-tests.samsungtv</li>
    <li>home-assistant-component-tests.tradfri</li>
    <li>python310Packages.adb-shell</li>
    <li>python310Packages.androidtv</li>
    <li>python310Packages.meshtastic</li>
    <li>python310Packages.pure-python-adb</li>
    <li>python310Packages.py3rijndael</li>
    <li>python310Packages.pygatt</li>
    <li>python310Packages.pytradfri</li>
    <li>python310Packages.samsungctl</li>
    <li>python310Packages.samsungtvws</li>
    <li>python39Packages.adb-shell</li>
    <li>python39Packages.androidtv</li>
    <li>python39Packages.meshtastic</li>
    <li>python39Packages.pure-python-adb</li>
    <li>python39Packages.py3rijndael</li>
    <li>python39Packages.pygatt</li>
    <li>python39Packages.pytradfri</li>
    <li>python39Packages.samsungctl</li>
    <li>python39Packages.samsungtvws</li>
  </ul>
</details>

dotlambda commented on pkgs/development/python-modules/pure-python-adb/default.nix:
```diff
@@ -11,11 +17,23 @@ buildPythonPackage rec {
+  ]
+  ++ passthru.extras-require.async;
+
+  pythonImportsCheck = lib.optionals doCheck [
+    "ppadb.client"
+    "ppadb.client_async"
+  ];
```
```suggestion
  pythonImportsCheck = [
    "ppadb.client"
  ] ++ lib.optionals doCheck [
    "ppadb.client_async"
  ];
```

mweinelt replied:
Again … good catch! Thank you.
